zookeeper-api/src/main/java/org/apache/helix/zookeeper/impl/factory/ZkConnectionManager.java

```diff
@@ -118,7 +118,7 @@ public void close() {
 
   protected synchronized void close(boolean skipIfWatched) {
     cleanupInactiveWatchers();
-    if (_sharedWatchers.size() > 0) {
+    if (_sharedWatchers != null && _sharedWatchers.size() > 0) {
```

Comment:
It seems `_shareWatchers` is initialized outside the constructor of `ZkConnectionManager`. If `ZkConnectionManager` completes initialization, `_sharedWatchers` should also complete initialization.
Doing the null check is unnecessary: an instance should be constructed well before it is used. `close()` should be called after the instance completes init. Why close() is called before the instance's state is fully initialized? We should find it out.

In one case I've known, it is related to SharedZkClient. The ZkConnection is created within sharedZkClient, before the ZK session is established, it is passed to ZkConnectionManager and ZkClient's connect(). And in ZkClient's connect, this is a check for this shared connection. Because the session is not established, zkclient has to `close()`. Meanwhile, `_shareWatchers` is not initialized.

I think we should fix the root cause, instead of covering the problem with this null check. what do you think?

Reply:
Synced offline. I missed the point ZkConnectionManager is managing connection. So there is a wait in ZkClient.
This NPE is thrown when ZkConnectionManager is being constructed and connection is not able to connect to ZK within timeout.